### DISCUSSION TITLE
Can a MultiResult contain another MultiResult?

### DISCUSSION THREAD

**Discussion by bytinbit** [nornir-automation/nornir#657] [Q&A] [answered]
I welcome the new type annotations with the latest 3.1.0 release of nornir. However, running mypy I stumbled upon this situation:

In a project I have the situation where the returned result types are as follows:
`task_results: MultiResult = [Result, MultiResult, MultiResult]`
and `task_results[1]` and `task_results[2]` contain elements of type `Result`. So it's a `MultiResult` that contains other `MultiResult` which in turn contain `Result`.

On the other hand, in nornir's `task.py`, `MultiResult` is constrained to contain type `Result`: 
```
class MultiResult(List[Result]):
 # .... code of class
```
However, nowhere in the documentation I find the definition how a `MultiResult` is constrained, so here's my question:

* Is a `MultiResult` allowed to contain further `MultiResult` elements? Then the type-annotation of the class is probably too constrained.
* Or can a `MultiResult` only contain `Result` types? Then I need to find a solution for this in my project and it would probably be a good idea to make that more explicit in nornir's documentation.

Thank you for the clarification and developing nornir, I hope to open-source my project soon (we're still in an early stage...).

**Comment by dbarrosop** [comment, marked as answer]
Well, technically speak a MultiResult is a list of Results by definition but you can have whatever you want inside a result. So you could have a result with the MultiResults inside.

Does that make sense?

An alternative would be to treat the various result objects as a Protocol instead, like we do with the plugins. That'd be the best way to solve the problem you are suggesting (which I think makes complete sense) but would require some significant effort.

**Reply by bytinbit** [reply]
Ah, so as I understand, a MultiResult contains Results, which in turn can contain MultiResult or Result or anything else.

I see if I can make it work in my code as you suggested (turn the MultiResult into a Result, which in turn contains other MultiResult). Maybe it would make sense to update the documentation to reflect this relationship/structure. Would it be welcomed if I made the appropriate note in the documentation with a PR? Possibly here: [Task Results: The programmatic way](https://nornir.readthedocs.io/en/latest/tutorial/task_results.html#The-programmatic-way)?

For the alternative: I'm yet too new to nornir to implement your Protocol idea on my own, let's hope that the longer I work with nornir, the more I'll be able to contribute...
